### PR TITLE
Partial fill rules

### PR DESCRIPTION
**Pull request by garbageslam**
A proposal to augment MCIP #31 to support "partial fill" atomic swap transactions

[Rendered Proposal](https://github.com/garbageslam/mcips/blob/partial-fill-rules/text/0042-partial-fill-rules.md)

### CODE REVIEW

**Comment by garbageslam** [comment]
The (current) proposal calls for adding `InputRuleVerificationData` separate from `InputRules` to `TxIn`. The idea is that `InputRules` is set by the originator and `InputRuleVerificationData` is set later by the counterparty who consumes the SCI. The `InputRuleVerificationData` is needed so that the counterparty has a place to put the amount shared secret for the "real" versions of each fractional input.

However, this has some drawbacks:
* The real versions of each fractional input must have the same amount shared secret as each fractional input (otherwise the originator won't be able to view-key match these outputs properly)
* Fortunately, these shared secrets already appear in the RevealedTxOut structure.
* The enclave would have to check that the shared secrets appearing in the InputRules match the shared secrets appearing in the verification data.

It is simpler to not have to duplicate these shared secrets into the verification data, and not have the verification data at all.

In the latest POC work we have done this, and will update the proposal soon.

**Comment by garbageslam** [comment]
An open question:

In MCIP #31 we proposed a structure "UnmaskedAmount" which is a value, token id, and blinding factor for a TxOut. This is part of the signed contingent input, so that the counterparty to a swap can build a Tx using an SCI. These data are deleted before they get to the enclave.

In this MCIP we proposed a different way to reveal the amount of a TxOut, via the amount shared secret.

The amount shared secret was introduced because it has some advantages over unmasked amount.

* The amount shared secret is verifiable, and we can confirm that the originator will be able to view-key match successfully. With the unmasked amount, it's possible that the counterparty can do something malicious and change the tx out shared secret used with the amount, so that it will still appear to unmask correctly, but the originator won't be able to decrypt it. This is critical for the partial fill application, and wasn't critical for the MCIP 31 application.
* The amount shared secret is shorter, only 32 bytes.

Since `UnmaskedAmount` represents an additional concept that we possibly don't need, it might make sense to retcon MCIP #31 to use amount shared secret rather than `UnmaskedAmount` in the SCI schemas.

File: text/0042-partial-fill-rules.md [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+     token id of the `partial_fill_output`.
+   * It must be the case that `fractional_output.value >= n/d * partial_fill_output.value` where `n` and `d` are the fill fraction
+     numerator and denominator. This equation must be validated by clearing denominators and checking an inequality of `u128`'s, to
+     avoid numerical issues.
+     
+     ```
+     fractional_output.value * fill_fraction_denominator >= fill_fraction_numerator * partial_fill_output.value
+     ```
+     
+     If this inequality does not hold, then the transaction must be rejected for reasons of not respecting the fill fraction.
```
Comment (nick-mobilecoin) [inline text/0042-partial-fill-rules.md:222]: :wrench: Even in the code I think this was a bit hard to follow.
I'm not sure it's much better, especially since it's already nested 2 levels.  Also it may be a misuse of **greifing**
```suggestion
   * It must be the case that:
      
     $$ {\texttt{partial_fill_change.value} - \texttt{fractional_change_output.value} \over \texttt{partial_fill_change.value}} <= {\texttt{fractional_output.value} \over \texttt{partial_fill_output.value}} $$
     
     **Note** Ideally we want these to be equal, but due to precision we favor the originator to prevent the counterparty from choosing values that allows griefing the originator.
     
     In the implementation the division is avoided, to further reduce precision loss:
     
     $$ ({\texttt{partial_fill_change.value} - \texttt{fractional_change_output.value}) \times \texttt{partial_fill_output.value}} <= {\texttt{fractional_output.value} \times \texttt{partial_fill_change.value}} $$
     
     If this inequality does not hold, then the transaction must be rejected for reasons of not respecting the fill fraction.
```

File: text/0042-partial-fill-rules.md [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This proposal has a few drawbacks.
+
+* The amounts of partial fill transactions are revealed to the enclave. Currently, MobileCoin transactions are validated using RingCT and do not reveal amounts to the enclave. It would be better if there were a way to validate partial fills without revealing this to the enclave, but for the use-cases envisioned, this data already has to be revealed anyways, so the use-cases aren't negatively impacted. We could imagine that an exchange service is based on SGX, and so the amounts of these quotes are only revealed inside of SGX, and are matched inside of SGX. Even in this case, it is still the case that if you penetrate SGX you will see the amounts of the SCI quotes.
```
Comment (nick-mobilecoin) [inline text/0042-partial-fill-rules.md:229]: :upside_down_face: : 
```suggestion
* The amounts of partial fill transactions are revealed to the enclave. Currently, MobileCoin transactions are validated using RingCT and do not reveal amounts to the enclave. It would be better if there were a way to validate partial fills without revealing this to the enclave, but for the use cases envisioned, this data already has to be revealed anyways, so the use cases aren't negatively impacted. We could imagine that an exchange service is based on SGX, and so the amounts of these quotes are only revealed inside of SGX, and are matched inside of SGX. Even in this case, it is still the case that if you penetrate SGX you will see the amounts of the SCI quotes.
```

File: text/0042-partial-fill-rules.md [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+# Motivation
+[motivation]: #motivation
+
+In [MCIP #31](https://github.com/mobilecoinfoundation/mcips/pull/0031) we proposed a way that two parties can swap assets -- the first one can make a signed offer to trade this for that using a "Signed Contingent Input", then anyone who agrees can add this to their transaction.
+
+The way it was proposed, the first party has to state the total amount of the first asset and
+the total amount of the second asset. But if the second party doesn't agree exactly, then they
+may not proceed, so there is a need to negotiate exactly the price and the volume.
+
+One way to make this more practical is that, the first party could instead sign that they are willing to trade at a particular price, up to a particular maximum volume. We call this a "partial fill", in analogy with partial-fill limit orders in traditional exchanges.
```
Comment (nick-mobilecoin) [inline text/0042-partial-fill-rules.md:20]: :wrench: it seems most places use "partial fill" as 2 words
```suggestion
One way to make this more practical is that, the first party could instead sign that they are willing to trade at a particular price, up to a particular maximum volume. We call this a "partial fill", in analogy with partial fill limit orders in traditional exchanges.
```

File: text/0042-partial-fill-rules.md [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+     
+     If this inequality does not hold, then the transaction must be rejected for reasons of not respecting the fill fraction.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This proposal has a few drawbacks.
+
+* The amounts of partial fill transactions are revealed to the enclave. Currently, MobileCoin transactions are validated using RingCT and do not reveal amounts to the enclave. It would be better if there were a way to validate partial fills without revealing this to the enclave, but for the use-cases envisioned, this data already has to be revealed anyways, so the use-cases aren't negatively impacted. We could imagine that an exchange service is based on SGX, and so the amounts of these quotes are only revealed inside of SGX, and are matched inside of SGX. Even in this case, it is still the case that if you penetrate SGX you will see the amounts of the SCI quotes.
+* Partial-fill SCI quotes cannot be re-used by a second counterparty later if the first counter-party only matched some of the volume.
```
Comment (nick-mobilecoin) [inline text/0042-partial-fill-rules.md:230]: :wrench: 
```suggestion
* Partial fill SCI quotes cannot be reused by a second counterparty later if the first counterparty only matched some of the volume.
```

File: text/0042-partial-fill-rules.md [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+* To verify the fractional change output, we need to verify an equation `(fraction_change - real_change) >= num / denom * real_change`.
+  This could similarly be verified without revealing commitments by clearing denominators and converting to the form `X >= 0`, then using
+  range proofs to verify the equation.
+
+The problem with this design is that, in MobileCoin, as with all RingCT-based cryptocurrencies, the amount of a TxOut actually appears in two places -- a Pedersen commitment, and a masked_value. The masked_value is a more conventional, symmetric-key encryption of the TxOut value, while the Pedersen commitment is an elliptic curve point which commits to the value, but cannot itself be directly "decrypted" efficiently by someone who doesn't already know the value.
+
+All recipients determine the value of their TxOut's by the following process:
+* Derive the TxOut shared secret
+* Use the TxOut shared secret to derive the value mask, token id mask, and the amount commitment blinding factor. This computation never fails with an error, but produces "conjectures" which are expected to be correct if the Recipient owns this TxOut.
```
Comment (nick-mobilecoin) [inline text/0042-partial-fill-rules.md:253]: :wrench: couple of suggestions
1. Don't think conjectures belongs in quotes
2. Don't think recipient should be title cased
```suggestion
* Use the TxOut shared secret to derive the value mask, token id mask, and the amount commitment blinding factor. This computation never fails with an error, but produces conjectures which are expected to be correct if the recipient owns this TxOut.
```

File: text/0042-partial-fill-rules.md [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+* Unmask the masked value and token id, obtaining a u64 value and a token id. These may or may not be the correct value of the Pedersen commitment.
+* Reconstruct the Pedersen commitment using these conjectured values and the amount commitment blinding factor. If this matches the commitment, then value recovery was successful.
+
+Thus value recovery relies on the masked_value and the amount commitment actually encoding the same numbers. If this is not the case, then value recovery fails, even if the TxOut is truly owned by the recipient.
+
+However, consensus does not validate in any way that the masked_value actually matches the amount commitment. It only validates the correctness of the commitments.
```
Comment (nick-mobilecoin) [inline text/0042-partial-fill-rules.md:259]: :wrench: 
```suggestion
However, consensus does not validate in any way that the masked value actually matches the amount commitment. It only validates the correctness of the commitments.
```

File: text/0042-partial-fill-rules.md [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+
+Historically, the rationale for this is as follows:
+* In a simple transaction, if the sender intentionally sends an output to a recipient that has a bad masked_value, this is equivalent to sending money to into a hole. The recipient will not be able to recover the value, so they will not render services to the sender.
```
Comment (nick-mobilecoin) [inline text/0042-partial-fill-rules.md:262]: :wrench: 
```suggestion
* In a simple transaction, if the sender intentionally sends an output to a recipient that has a bad masked value, this is equivalent to sending money to into a hole. The recipient will not be able to recover the value, so they will not render services to the sender.
```

File: text/0042-partial-fill-rules.md [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+* In MCIP #31 atomic swaps without partial fills, the originator is setting the masked_value for the required outputs which are going to them. If the counterparty changes the masked_value at all, the transaction is invalid. So the originator can only poison their own masked value, which only throws their own money into a hole.
+
+(This is also the reasoning for all other RingCT-based cryptocurrencies.)
+
+However, once we start doing partial fills, now we have a situation where the counter-party has to create the masked value for the originator. If consensus does not in some way validate the masked value, then a griefing attack becomes possible:
```
Comment (nick-mobilecoin) [inline text/0042-partial-fill-rules.md:267]: :upside_down_face: 
```suggestion
However, once we start doing partial fills, now we have a situation where the counterparty has to create the masked value for the originator. If consensus does not in some way validate the masked value, then a griefing attack becomes possible:
```

File: text/0042-partial-fill-rules.md [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+
+However, this is an extremely complicated solution to this problem. Embedding both Ristretto scalar multiplication and blake2b hashing into a ZK-SNARK would mean that we are building zero-knowledge proofs that are similar in complexity to Zcash. Making this mobile-friendly becomes a challenging research problem. So this is for now a dead-end and something we could only hope to do later as part of an upgrade to all of the transaction math.
+
+Another approach might involve migrating the masked_value from a simple symmetric-key encryption to something with homomorphic properties, so that we can verify it similarly to how we verify the partial-fill equations for the Pedersen commitments. However, this ultimately doesn't seem very promising, because even if we could homomorphically verify inequalities over the masked values, that isn't enough, because the masked value for each output has to actually match exactly to the corresponding Pedersen commitment, and two inequalities being satisfied doesn't imply this.
```
Comment (nick-mobilecoin) [inline text/0042-partial-fill-rules.md:277]: :wrench: 
```suggestion
Another approach might involve migrating the masked value from a simple symmetric-key encryption to something with homomorphic properties, so that we can verify it similarly to how we verify the partial fill equations for the Pedersen commitments. However, this ultimately doesn't seem very promising, because even if we could homomorphically verify inequalities over the masked values, that isn't enough, because the masked value for each output has to actually match exactly to the corresponding Pedersen commitment, and two inequalities being satisfied doesn't imply this.
```

File: text/0042-partial-fill-rules.md [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+None at this time.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+In the future, if/when we replace SGX transaction validation entirely with zero-knowledge proofs, we should use that proving framework to also
+support zero knowledge partial fill validation.
```
Comment (nick-mobilecoin) [inline text/0042-partial-fill-rules.md:296]: :wrench: 
```suggestion
support zero-knowledge partial fill validation.
```

**Review by nick-mobilecoin — APPROVED** [review]

File: text/0042-partial-fill-rules.md [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+
+# Summary
+[summary]: #summary
+
+Extend the set of supported "input rules", to allow that someone can produce a signed quote indicating willingness to trade "this for that", but moreover, "this for that, at this exchange rate, up to this volume".
+
+# Motivation
+[motivation]: #motivation
+
+In [MCIP #31](https://github.com/mobilecoinfoundation/mcips/pull/0031) we proposed a way that two parties can swap assets -- the first one can make a signed offer to trade this for that using a "Signed Contingent Input", then anyone who agrees can add this to their transaction.
```
[the remark this answers is not in the document]
Comment (garbageslam) [reply text/0042-partial-fill-rules.md:14]: I am not sure, I've generally been linking to the PR

@jcape do you have an opinion on this?

File: text/0042-partial-fill-rules.md [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+
+The way it was proposed, the first party has to state the total amount of the first asset and
+the total amount of the second asset. But if the second party doesn't agree exactly, then they
+may not proceed, so there is a need to negotiate exactly the price and the volume.
+
+One way to make this more practical is that, the first party could instead sign that they are willing to trade at a particular price, up to a particular maximum volume. We call this a "partial fill", in analogy with partial-fill limit orders in traditional exchanges.
+
+**Note** however, that this is not conceptually an implementation of an "exchange" which accepts and then determines settlement of orders. It's more like a fancy quoting service, wherein one party can offer a signed quote which another party may act on or ignore.
+
+We want to extend the [MCIP #31](https://github.com/mobilecoinfoundation/mcips/pull/0042) functionality to support this.
```
[the remark this answers is not in the document]
Comment (garbageslam) [reply text/0042-partial-fill-rules.md:24]: thank you!

File: text/0042-partial-fill-rules.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (garbageslam) [reply text/0042-partial-fill-rules.md:14]: I've happy to link to the text instead, I guess that makes more sense for md files that become part of the set of mcips

File: text/0042-partial-fill-rules.md [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+
+An example partial fill flow goes like this:
+* Alice (the originator) has an output for 1500 MOB. Alice would like to trade up to 1000 MOB for MEOWB at a price of 10 MEOWB/MOB.
+  Alice doesn't want the trade to happen unless the counterparty takes at least .1 MOB.
+* Alice creates an SCI using her 1500 MOB input.
+  * She creates a required output to herself for 500 MOB
+  * She creates a fractional output to herself for 10,000 MEOWB
+  * She creates a fractional change output to herself for 1000 MOB
+  * She sets the min_fill_value to .1 MOB, (but expresses this in picoMOB, the smallest representable units).
```
[the remark this answers is not in the document]
Comment (garbageslam) [reply text/0042-partial-fill-rules.md:53]: thanks -- I think I should have replaced this with `min_partial_fill_value` since that's what we named this field in the end.

File: text/0042-partial-fill-rules.md [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+When the transaction settles, Bob has paid net 200 MEOWB + transaction fee, and received net of 20 MOB.
+Alice has paid a 1500 MOB output, and received the 500 MOB required output, and a 980 MOB fractional output,
+as well as 200 MEOWB from Bob.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The reference level explanation splits into two parts -- the Amount shared secret, and the schema and validation of the new SCI fields.
+
+Neither of these are legal before block version 3. The new SCI fields are legal in block version 3, and the alternate derivation for
```
[the remark this answers is not in the document]
Comment (garbageslam) [reply text/0042-partial-fill-rules.md:69]: Maybe we should update the MCIP 26 table? https://github.com/mobilecoinfoundation/mcips/blob/main/text/0026-block-version-based-protocol-evolution.md

I tend to agree that we should have a block version changelog, maybe in a separate file at the root of this repo, that doesn't have all the extra content in MCIP 26

[the remark this answers is not in the document]
Comment (garbageslam) [reply text/0042-partial-fill-rules.md:69]: @jcape what do you think about adding a file like `text/block_version_features.md` which contains a list of block versions, short description of what's new in them, and link to relevant MCIPs?

File: text/0042-partial-fill-rules.md [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+
+Recall that the `InputRules` are created (and signed) by the **originator**, and this TxIn is added to a transaction
+by the **counterparty**, who is also obligated to add TxOut's to the TxPrefix which match to the partial fill change
+and partial fill outputs appropriately. (This is actually done under the hood by the transaction builder.)
+
+The consensus enclave must enforce any input rules, including these new partial fill rules. This is done in the following way.
+
+1. If `partial_fill_change` is not present, then no other partial fill rules may be used, and any other partial fill fields must be defaulted.
+1. If `partial_fill_change` is present, then decrypting the amount and token id using `amount_shared_secret` must succeed. The partial fill change
+   may not have a 0 value (to help clients avoid division by zero).
```
[the remark this answers is not in the document]
Comment (garbageslam) [reply text/0042-partial-fill-rules.md:200]: You're right, we have this check in the input rules code:

https://github.com/mobilecoinfoundation/mobilecoin/blob/117e2c27900e9b22514d5eb69401391369cf1240/transaction/core/src/input_rules.rs#L104

I guess what's interesting here is, the enclave doesn't actually do any division, since it multiplies things out. So the enclave doesn't actually care itself right now if the partial fill change is zero.

I think what we should do is:

* Insert a check for the partial fill change being zero into the input rules verification. Even if it's not really required to protect the enclave, it might be simpler to explain the spec if we just enforce this everywhere.
* At line 104 we should insert a code comment that this prevents a later subtraction expression from panicking.

[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply text/0042-partial-fill-rules.md:200]: Checking in the enclave for consistency sounds good.

Just to double check, is the panic at [138](https://github.com/mobilecoinfoundation/mobilecoin/blob/master/transaction/core/src/input_rules.rs#L138) the only one that prevents creating mob or is there still a subsequent check with the RingCT logic (that I still need to find time to get a high level understanding of)

File: text/0042-partial-fill-rules.md [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+RevealedTxOut --> InputRules
+
+TxIn: Vec~TxOut~ ring
+TxIn: Vec~...~ membership_proofs
+TxIn: Option~InputRules~ input_rules
+
+InputRules: Vec~TxOut~ required_outputs
+InputRules: uint64 max_tombstone_block
+InputRules: Vec~RevealedTxOut~ partial_fill_outputs *new*
+InputRules: RevealedTxOut partial_fill_change *new*
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply text/0042-partial-fill-rules.md]: I think the downside with protobuf is aren't things _optional_ by default?  Perhaps "required" should be part of the description...???

File: text/0042-partial-fill-rules.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (garbageslam) [reply text/0042-partial-fill-rules.md:200]: The RingCT logic ensures that every transaction is "balanced" -- the value of every output in tx prefix + the fee output = the value of every true input. That check happens in `transaction/core/src/ring_ct/signature_rct_bulletproofs.rs` in the verify function somewhere

File: text/0042-partial-fill-rules.md [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+* The same input can be used to "back" multiple SCI quotes at once, in which case if any one of them is matched, the others instantly become impossible to match. This is somewhat different from how limit orders work in an exchange -- each limit order must be backed by distinct liquidity. This reflects the idea that SCI's are more in analogy with a signed quote than with a limit order as such. This is potentially confusing to e.g. a trader who looks at a list of available SCI quotes and infers that there is a certain amount of liquidity depth, reasoning as they would in a traditional exchange. But if the same inputs are used to sign multiple quotes e.g. for different currencies, it may be impossible that all of these quotes could actually match, and the "real" liquidity may be considerably less. This can be mitigated by making sure that the software they are using communicates this idea clearly.
```
[the remark this answers is not in the document]
Comment (garbageslam) [reply text/0042-partial-fill-rules.md:235]: > What ensures this, is it the key image specifically tied to the txout.

Yes -- each TxOut has a unique KeyImage. This is revealed when an MLSAG spending that TxOut is signed, and there is no way to make a valid MLSAG signature that reveals something other than that one true key image. This is the mechanism that prevents double spends in MobileCoin.

An SCI, containing one MLSAG, also has a key image. Nothing stops you from signing more than one MLSAG or SCI over the same TxOut, but they all become mutually exclusive because transaction validation prevents a key image from appearing twice in the ledger.

> If someone doesn't have a full amount would they first need to make a tx that combines their chunks into one?

Not necessarily I think? If they have 3 inputs, they could sign 3 different SCIs all with the same implied price? They could consolidate the Tx's first

> Is there a crypto term for combining, smelt..??

I am not sure if there's a standard term, I've used the term "consolidate", we also have a concept of "wallet fragmentation" errors and a "defragmentation Tx" to fix it. There's a common term "dust" that I know of.

:laughing: 
Whoever smelt it dealt it

File: text/0042-partial-fill-rules.md [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+1. If `partial_fill_change` is not present, then no other partial fill rules may be used, and any other partial fill fields must be defaulted.
+1. If `partial_fill_change` is present, then decrypting the amount and token id using `amount_shared_secret` must succeed. The partial fill change
+   may not have a 0 value (to help clients avoid division by zero).
+1. If `partial_fill_change` is present, then a corresponding TxOut called the `fractional_change_output` must exist in the `TxPrefix.outputs`.
+   This must match the `partial_fill_change` in every field except possibly the `masked_amount`.
+1. The `fractional_change_output` must decrypt successfully using the `amount_shared_secret` of the `partial_fill_change`, and
+   it's amount must have the same token id as the `partial_fill_change`, and its value must be less or equal.
+1. The fill fraction is inferred at this point.
+   * The numerator is `partial_fill_change.value - fractional_change_output.value`.
+   * The denominator is `partial_fill_change.value`.
```
[the remark this answers is not in the document]
Comment (garbageslam) [reply text/0042-partial-fill-rules.md:207]: I would rather not use latex in this doc, but interesting suggestion

File: text/0042-partial-fill-rules.md (hunk 1 above)
Comment (garbageslam) [reply text/0042-partial-fill-rules.md]: I largely took this change, except for the latex part

**Comment by garbageslam** [comment]
[Rendered block-version-features registry](https://github.com/garbageslam/mcips/blob/partial-fill-rules/registries/block-version-features.md)

File: text/0042-partial-fill-rules.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply text/0042-partial-fill-rules.md:247]: > Whoever smelt it dealt it

:)

**Comment by nick-mobilecoin** [comment]
1. I think we need some kind of block version features.  The proposed [block-version-features.md](https://github.com/garbageslam/mcips/blob/partial-fill-rules/registries/block-version-features.md) is a good approach.

2. As for latex, I think that suggestion was probably overkill for this specific PR. I was probably guilty of seeing a shiny new tool to play with.  Some downsides to latex:
   - The raw markup is not intuitive to understand
   - It leads to it's own debugging issues as it's more or less code in markdown
  
  Especially for this PR the only benefit would have been _nicer_ fractions and IFF that was desired it could be done in ascii art.
  ```
   partial_fill_change.value - fractional_change_output.value         fractional_output.value       
  ------------------------------------------------------------  ==  ----------------------------
                    partial_fill_change.value                        partial_fill_output.value
  ```
  Versus
  
  $$ {\texttt{partial fill change.value} - \texttt{fractional change output.value} \over \texttt{partial fill change.value}} == {\texttt{fractional output.value} \over \texttt{partial fill output.value}}$$

**Comment by nick-mobilecoin** [comment]
While Latex is of questionable value, I think mermaid use is very valuable.  

A diagram can communicate so much in so little.  Often it is one of the best mediums to communicate relationships between objects, components, etc.
Using ASCII art for diagrams is a big pain to write.  Updating ASCII art diagrams is just as much of a pain.

**Comment by garbageslam** [comment]
Yeah, I feel like, having a registry of the features can help give us a source of truth for all this stuff:

https://github.com/mobilecoinfoundation/mobilecoin/blob/631bdb443ca1338836431af0f01ea3c30f006ec1/transaction/types/src/block_version.rs#L79

File: text/0042-partial-fill-rules.md [hunk 19]
```diff
@@ -0,0 +1,316 @@
...
+    A[TxOut SharedSecret] -->|blake2b| D[Amount Blinding Factor]
+    A[TxOut SharedSecret] -->|blake2b| E[Value mask]
+```
+
+After:
+```mermaid
+graph TB
+    A[TxOut SharedSecret] -->|hkdf sha512| B[MemoKey]
+    A -->|blake2b| C[Confirmation Number]
+    A -->|sha512| D[Amount Shared Secret *new*]
```
Comment (wjuan-mob) [inline text/0042-partial-fill-rules.md:95]: Do you mind adding the explanation here for why we transitioned from blake2b to sha512?

**Review by wjuan-mob — APPROVED** [review]

File: text/0042-partial-fill-rules.md [hunk 20]
```diff
@@ -0,0 +1,316 @@
...
+
+An example partial fill flow goes like this:
+* Alice (the originator) has an output for 1500 MOB. Alice would like to trade up to 1000 MOB for MEOWB at a price of 10 MEOWB/MOB.
+  Alice doesn't want the trade to happen unless the counterparty takes at least 0.1 MOB.
+* Alice creates an SCI using her 1500 MOB input.
+  * She creates a required output to herself for 500 MOB
+  * She creates a fractional output to herself for 10,000 MEOWB
```
[the remark this answers is not in the document]
Comment (garbageslam) [reply text/0042-partial-fill-rules.md:51]: yes, thank you, i failed to update this properly when we changed terminology

File: text/0042-partial-fill-rules.md (hunk 19 above)
Comment (garbageslam) [reply text/0042-partial-fill-rules.md:95]: I think the reasoning was like:

* `HKDF<Blake2b>` is less well-studied than `HKDF<SHA512>` and the recommendation was not to use `HKDF<Blake2b>` anywhere.
* At one point, `TxOutSharedSecret -> AmountSharedSecret` was an HKDF derivation. However, since there was only one thing to expand, and HKDF involves a second invocation of `SHA512`, it seemed simpler to skip the second `SHA512`, and just use `SHA512` once at this step.

Arguably, we should just use `Blake2b` anywhere are using a hash function a single time, and only use `SHA512` when we have to use `HKDF`. So maybe this should just be changed to `Blake2b` here.

Good question, thanks. Do you have an opinion about which way we should go?

Comment (garbageslam) [reply text/0042-partial-fill-rules.md:98]: Following off-line discussion, we want to make this `Blake2b`, at least to be more consistent with everything else. The only time we need to prefer `Sha512` is when using `HKDF`.

**Review by jcape — APPROVED** [review]
LGTM, albeit I'd avoid pre-supposing the existence of a quoting service since it's not strictly required for this.
